src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java

```diff
@@ -52,11 +58,35 @@ private CredentialsFactory() {
                 return Optional.of(new AwsCertificateCredentials(name, description, new SecretBytesSupplier(client, name)));
             case Type.file:
                 return Optional.of(new AwsFileCredentials(name, description, filename, new SecretBytesSupplier(client, name)));
+            case Type.githubApp:
+                return GitCredentialFactory.createCredential(name, description, appId, Secret.fromString(new StringSupplier(client, name).get()));
             default:
                 return Optional.empty();
         }
     }
 
+    public static class GitCredentialFactory  {
+
+        private static final Logger LOG = Logger.getLogger(GitCredentialFactory.class.getName());
+
+        public static Optional<StandardCredentials> createCredential(String name, String description, String appId, Secret secret) {
+            if (Jenkins.get().getPlugin("github-branch-source") == null) {
+                LOG.log(Level.WARNING, "Plugin not installed: github-branch-source. Cannot create type: " + Type.githubApp);
+                return Optional.empty();
+            }
+
+            try {
+                Class githubCredentials = Class.forName("org.jenkinsci.plugins.github_branch_source.GitHubAppCredentials");
```

Comment:
Could you point to other examples of credentials being constructed using reflection, if there are examples out there? Obviously anything using reflection warrants scrutiny, so I'd feel more comfortable about this if I can see that a pattern is being followed.

Reply:
Hi Chris, I wanted to use reflection to have a weak dependency on the github-branch-source plugin and avoid forcing the user to install two plugins. Here is an example of another plugin interacting with github-branch-source

[pipeline-model-definition-plugin](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/master/pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/when/impl/DefaultChangeLogStrategy.java)

Reply:
I remembered that Jenkins also has an idiomatic way of using optional dependencies in code the straightforward way - without reflection - where you put the `@Extension(optional = true)` annotation on the classes that use the optional dependencies. This suppresses the ClassNotFound errors when that class is loaded and the dependency plugin is not present on the system. If you can find the right place to add this annotation I think it would be an improvement on the reflection approach.

Example here: https://github.com/jenkinsci/aws-secrets-manager-credentials-provider-plugin/commit/06ad79e458d92368d599d5fa1cc1bbd088f46e7b#diff-2a94427d0a60f95532b13490458755d1ef9201c662e932df503ac6014f1fd3a9R27

Reply:
Ok, I moved the GithubCredentialsFactory into it a new class that uses @Extension(optional = true). I Also created a test that mimics StandardUsernamePasswordCredentialsIT. I can't test the password retrieval because the GithbAppCredentials uses api.github.com/app inside the method, because I was able to add tests for the functionality that doesn't use the password.